datadog/resource_datadog_role.go

```diff
@@ -0,0 +1,288 @@
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/DataDog/datadog-api-client-go/api/v2/datadog"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/customdiff"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+// validPermissions is a map of all unrestricted permission IDs to their name
+var validPermissions map[string]string
+
+func resourceDatadogRole() *schema.Resource {
+	return &schema.Resource{
+		Exists: resourceDatadogRoleExists,
+		Create: resourceDatadogRoleCreate,
+		Read:   resourceDatadogRoleRead,
+		Update: resourceDatadogRoleUpdate,
+		Delete: resourceDatadogRoleDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceDatadogRoleImport,
+		},
+		CustomizeDiff: customdiff.ValidateValue("permission", validatePermissionsUnrestricted),
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of the role.",
+			},
+			"permission": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Set of objects containing the permission ID and the name of the permissions granted to this role.",
+				Elem:        getRolePermissionSchema(),
+			},
+			"user_count": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "Number of users that have this role.",
+			},
+		},
+	}
+}
+
+func getValidPermissions(client *datadog.APIClient, auth context.Context) (map[string]string, error) {
+	// Get a list of all permissions, to ignore restricted perms
+	if validPermissions == nil {
+		res, _, err := client.RolesApi.ListPermissions(auth).Execute()
+		if err != nil {
+			return nil, translateClientError(err, "error listing permissions")
+		}
+		permsList := res.GetData()
+		permsNameToId := make(map[string]string, len(permsList))
+		for _, perm := range permsList {
+			if !perm.Attributes.GetRestricted() {
+				permsNameToId[perm.GetId()] = perm.Attributes.GetName()
+			}
+		}
+		validPermissions = permsNameToId
+	}
+	return validPermissions, nil
+}
+
+func validatePermissionsUnrestricted(value interface{}, meta interface{}) error {
+	client := meta.(*ProviderConfiguration).DatadogClientV2
+	auth := meta.(*ProviderConfiguration).AuthV2
+
+	// Get a list of all valid permissions
+	validPerms, err := getValidPermissions(client, auth)
+	if err != nil {
+		return err
+	}
+
+	perms := value.(*schema.Set)
+	for _, permI := range perms.List() {
+		perm := permI.(map[string]interface{})
+		permID := perm["id"].(string)
+		if _, ok := validPerms[permID]; !ok {
+			return fmt.Errorf(
+				"permission with ID %s is restricted and cannot be managed by terraform or does not exist, remove it from your configuration",
+				permID,
+			)
+		}
+	}
+
+	return nil
+}
+
+func getRolePermissionSchema() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "ID of the permission to assign.",
+			},
+			"name": {
```

Comment:
Is this mostly here for the ability to reference?

Reply:
Yeah. I'm not sure about this one though. At first I had a simple string list of IDs, then I had to add it for my first try at implanting `diffSupressFunc` that i didn't end up doing, but kept this as is 🤷